DOC-756: Document new table row numbering (series column) advtable feature (#2086)
tinymce/tinymce-docs: _includes/configuration/advtable_value_series.md

## `advtable_value_series`
{{ site.requires_5_9v }}

> {% if pluginname != "Advanced Tables" %}

Reviewer: Why do we have this? I can't see this being included anywhere else in this PR.

Author: I just kept it from the template file. I'll remove it since this wouldn't be used outside of the advtable page. Thanks.

> **Note**: The `advtable_value_series` option requires the Advanced Tables plugin.
{% endif %}

The `advtable_value_series` option configures value series for populating cells in a table. This option can be used to create row identifiers.

**Type:** `Object`

**Default Value:**

```js
{
  // Natural number series
  numeric: { 
    title: 'Numeric',
    update: true,
    resizable: false,
    generator: `GeneratorFunction` // For details, see: 'Creating a value series generator'
  },
  // English alphabetic series
  alpha: {
    title: 'Alpha',
    update: true,
    resizable: false,
    generator: `GeneratorFunction` // For details, see: 'Creating a value series generator'
  },
}
```

Both default series are configured to update on table changes and not resize when using the resize bars.

> ### Series configuration

Reviewer 2: Should we explain that the "name" of the series column is the key for this object?

Author: Good point. I'll try and draft something up.

Author: Does something like this seem suitable?
`Each key in the 'advtable_value_series' option object indicates the name of the value series and its configuration.`

Author: Have added something very similar above the 'Series configuration' heading. Hopefully, it is clearer now.

| Name | Value | Requirement | Description |
| ---- | ----- | ----------- | ----------- |
| title | `string` | Required | The name shown in the UI for the series. |
| update | `boolean` | Optional | default: `false` - When `true`, the series values will be updated when changes are made to the table. |

> | resizable | `boolean` | Optional | default: `true` - When `true`, table cells containing the series values can be resized using a mouse or touch device. |

Reviewer 2: Wondering whether we need to specify the input types

```suggestion
| resizable | `boolean` | Optional | default: `true` - When `true`, table cells containing the series values can be resized. |
```

Author: I think the input types was something Tyler had put in one of the suggestions in the original PR. I don't have any issues with keeping or removing it personally.

Reviewer: My preference would be to remove it as well, but I don't have any strong opinions about it.

Reviewer 3: let's leave them in for clarity.

| generator | `(info: GeneratorInfo, rowIndex: number, columnIndex: number) => GeneratorResult` | Required | For details on creating a value series generator, see: [Creating a value series generator](#creatingavalueseriesgenerator).  |

#### Creating a value series generator

> The `generator` is a callback function for specifying how to update a table cell of a value series. The callback is passed information relating to the generator and table cell, the row index, and column index of the table cell. For details, see: [GeneratorInfo](#generatorinfo). The callback should return an object containing the value and optionally, any classes and attributes to be applied to the table cell. For details, see: [GeneratorResult](#generatorresult).

Reviewer 2: These tweaks are just my opinion

```suggestion
The `generator` is a callback function used to specify how a table cell of a value series will update. The callback is passed information relating to the generator and table cell, the row index, and column index of the table cell. For details, see: [GeneratorInfo](#generatorinfo). The callback should return an object containing the value and optionally, any classes and attributes to be applied to the table cell. For details, see: [GeneratorResult](#generatorresult).
```

If state needs to be kept between generator iterations, additional properties can be added to the generator result. The state can be accessed through the `prev` property of the `info` parameter. For details, see: [GeneratorInfo](#generatorinfo). 

##### GeneratorInfo

| Name | Value | Description |
| ---- | ----- | ----------- |
| sectionType | `'thead'`, `'tbody'` or `'tfoot'` | The section of the table cell. |
| cellType | `'td'` or `'th'` | The type of the table cell. |
| classes | `string[]` | The classes present on the table cell. |
| direction | `'row'` or `'column'` | The direction of the generator. |
| prev | `GeneratorResult` | The generator result from the previous iteration. |

##### GeneratorResult

> 

Author: Thanks for the suggestion.

> ## `advtable_value_series`
> {{ site.requires_5_9v }}

Reviewer: ```suggestion
## `advtable_value_series`

{{ site.requires_5_9v }}
```
I'm not too sure if this would cause issues, but elsewhere we keep a blank line between these to make sure they don't "merge"

> The `advtable_value_series` option configures value series for populating cells in a table. This option can be used to create row identifiers.

Reviewer: ```suggestion
The `advtable_value_series` option configures a series of values for populating cells in a table. This option can be used to create row identifiers.
```
Maybe?

Author: I think my only concern with that is that it could sound like only one value series can be created. Maybe being more explicit would be better?
```suggestion
The `advtable_value_series` option configures one or more value series for populating cells in a table. This option can be used to create row identifiers.
```

Reviewer 3: ```suggestion
The `advtable_value_series` option configures one or more series of values for populating cells in a table. This option can be used to create row identifiers.
```
Split the difference?

## `advtable_value_series`

{{ site.requires_5_9v }}

The `advtable_value_series` option configures value series for populating cells in a table. This option can be used to create row identifiers.

**Type:** `Object`

**Default Value:**

```js
{
  // Natural number series
  numeric: { 
    title: 'Numeric',
    update: true,
    resizable: false,
    generator: `GeneratorFunction` // For details, see: 'Creating a value series generator'
  },
  // English alphabetic series
  alpha: {
    title: 'Alpha',
    update: true,
    resizable: false,
    generator: `GeneratorFunction` // For details, see: 'Creating a value series generator'
  },
}
```

Both default series are configured to update on table changes and not resize when using the resize bars.

> Each property of the `advtable_value_series` object indicates the name of the value series and its configuration.

Reviewer 3: Each top-level properties of the `advtable_value_series` object are used as the name of the value series and its configuration. In the following example, there are two value series named "`numeric`" and "`alpha`":

```js
{
  numeric: { 
    title: 'Numeric',
    update: true,
    resizable: false,
    generator: `GeneratorFunction`
  },
  alpha: {
    title: 'Alpha',
    update: true,
    resizable: false,
    generator: `GeneratorFunction`
  },
}
```

> Each property of the `advtable_value_series` object indicates the name of the value series and its configuration.

Reviewer 3: Maybe rewrite like the comment above (I can't remember how to codeblock in codeblocks for github suggestions)

> Each property of the `advtable_value_series` object indicates the name of the value series and its configuration.

Reviewer 3: Fixed in https://github.com/tinymce/tinymce-docs/pull/2086/commits/31168e42dea8df8363aba80c5ee96430dd911656

### Series configuration

> | Name | Value | Requirement | Description |

Reviewer 3: should the "Name" field be something else?

| ---- | ----- | ----------- | ----------- |
| title | `string` | Required | The text shown in the UI for the series. |
| update | `boolean` | Optional | default: `false` - When `true`, the series values will be updated when changes are made to the table. |
| resizable | `boolean` | Optional | default: `true` - When `true`, table cells containing the series values can be resized using a mouse or touch device. |
| generator | `(info: GeneratorInfo, rowIndex: number, columnIndex: number) => GeneratorResult` | Required | For details on creating a value series generator, see: [Creating a value series generator](#creatingavalueseriesgenerator).  |

#### Creating a value series generator

> The `generator` is a callback function used to specify how a table cell of a value series will update. The callback is passed information relating to the generator and table cell, the row index, and column index of the table cell. For details, see: [GeneratorInfo](#generatorinfo). The callback should return an object containing the value and optionally, any classes and attributes to be applied to the table cell. For details, see: [GeneratorResult](#generatorresult).

Reviewer 3: ```suggestion
The `generator` is a callback function used to specify how a table cell of a value series will update. The callback is passed information relating to: the generator and table cell, the row index, and column index of the table cell. For details, see: [GeneratorInfo](#generatorinfo). The callback should return an object containing the value and optionally, any classes and attributes to be applied to the table cell. For details, see: [GeneratorResult](#generatorresult).
```
?

> If state needs to be kept between generator iterations, additional properties can be added to the generator result. The state can be accessed through the `prev` property of the `info` parameter. For details, see: [GeneratorInfo](#generatorinfo). 

Reviewer 3: ```suggestion
If the "state" of the series needs to be kept between generator iterations, additional properties can be added to the generator result. The state can be accessed through the `prev` property of the `info` parameter. For details, see: [GeneratorInfo](#generatorinfo). 
```
?

##### GeneratorInfo

An object with the following properties is passed to the generator callback function as the `info` parameter.

| Name | Value | Description |
| ---- | ----- | ----------- |
| sectionType | `'thead'`, `'tbody'` or `'tfoot'` | The section of the table cell. |
| cellType | `'td'` or `'th'` | The type of the table cell. |
| classes | `string[]` | The classes present on the table cell. |
| direction | `'row'` or `'column'` | The direction of the generator. |
| prev | `GeneratorResult` | The generator result from the previous iteration. |

##### GeneratorResult

The generator callback function should return an object with the following properties.

| Name | Value | Requirement | Description |
| ---- | ----- | ----------- | ----------- |
| classes | `string[]` | Optional | The classes to be applied to the table cell. |

> | attributes | `Object` | Optional | The attributes to be applied to the table cell. The `attributes` should be provided as an object where each key is an attribute and each value is of type `string`, `boolean`, `number` or `null`. A value of `null` for an attribute will remove the attribute from the table cell. |

Reviewer 3: ```suggestion
| attributes | `Object` | Optional | The attributes to be applied to the table cell. The `attributes` should be provided as an object where each key is an attribute and each value is of type `string`, `boolean`, `number`, or `null`. A value of `null` for an attribute will remove the attribute from the table cell. |
```
